add Functors docs (#1654)
FluxML/Flux.jl: docs/src/models/functors.md

## Functors.jl

> Flux makes use of the Functors.jl to represent many of the core functionalities it provides.

Reviewer: ```suggestion
Flux makes use of the Functors.jl to perform recursive introspection into layers and models.
```

> Flux makes use of the Functors.jl to represent many of the core functionalities it provides.

Reviewer 2: ```suggestion
Functors.jl allows for recursive traversal of deeply nested structures, making it possible to treat a complex struct like any other collection in Julia.

It is used extensively throughout [Flux.jl](https://github.com/FluxML/Flux.jl).
```

Author: If we want to give details about Functors.jl I would rather say that it is a set of utility functions designed to help with treating certain safe structs as Functors. Notably ones include most of the layers that Flux defines.

Traversal/ nesting is strictly orthogonal.

Reviewer 2: Okay then say that, but we should say what Functors.jl is instead of just that another package uses it.

Reviewer 3: Alternatively, give Functors its own standalone docs site like Zygote and only talk about the Flux relevant parts here. After all, you can use both it and dependents like Optimisers without ever touching Flux.

Reviewer 2: ....wow I totally thought this was a PR to Functors.jl this entire time

Reviewer 2: (and I agree that Functors.jl should have its own docs site)

You can ignore my comments here.

Author: https://github.com/FluxML/Functors.jl/pull/17  does that too

```@docs
Functors.isleaf
Functors.children
Functors.fcollect

> Functors.fmap

Reviewer 2: ```suggestion
Functors.fmap
Functors.fmapstructure
```

Author: I am coming around to the notion that `fmapstructure` isn't a real API at all. If anything it can be put in the utils (with a different name) here and be removed from Functors.

Reviewer 2: Then remove it in another PR, but if it's exported and has a docstring, then it should be in the docs.
